Add initial forms components

tomas-sucena opened the pull request:
As discussed in the previous sprint, @Lauranea and I started developing form components. So far, we have:

| Component | Developed by | Screenshot |
|:-------------:|:----------------:|:------------:|
| Picture Input | @Lauranea | ![image](https://github.com/user-attachments/assets/7fb79fff-05df-4a5c-afda-87347bc73ea7) |
| Radio Buttons | @tomas-sucena (me) |  ![image](https://github.com/user-attachments/assets/a4c6db7d-d941-448f-a449-550729f621dc) |

We've also added a new folder - `forms/` - inside `src/components/`  to store these and all future forms components.
 
Please feel free to modify our code to your liking. Since we're just starting out, we'd appreciate any feedback we could get. 😄  

# Review checklist

- [ ] Contains enough appropriate tests
- [ ] Behavior is as expected
- [ ] Clean, well-structured code

tomas-sucena committed:
feat: first radio-button draft

Using CSS only.

Co-Authored-By: Sofia Minnemann <93668178+Lauranea@users.noreply.github.com>

Lauranea committed:
feat: start upload picture

Lauranea committed:
added more padding

tomas-sucena committed:
refactor: style radio buttons

Replace most CSS with Tailwind.

tomas-sucena committed:
fix: change display of uploaded image

tomas-sucena committed:
chore: add radio buttons to storybook

Lauranea committed:
start of upload multiple files

tomas-sucena committed:
fix(radio button): fix parameter type

tomas-sucena committed:
chore(storybook): add PictureInput story

tomas-sucena committed:
chore: format code

MRita443 commented on src/lib/components/forms/picture-input.stories.ts:
```diff
@@ -0,0 +1,18 @@
+  }
+};
+
+export const Example = {
```
```suggestion
  title: 'Atoms/Forms',
  component: PictureInput,
  argTypes: {
    text: { control: 'text' }
  },
  parameters: {
    layout: 'centered'
  }
};

export const Picture_Input = {
```

MRita443 commented on src/lib/components/forms/radio-buttons.stories.ts:
```diff
@@ -0,0 +1,18 @@
+  }
+};
+
+export const Example = {
```
```suggestion
  title: 'Atoms/Forms',
  component: RadioButtons,
  argTypes: {
    options: { control: 'array' }
  },
  parameters: {
    layout: 'centered'
  }
};

export const Radio_Buttons = {
```

MRita443 commented on src/lib/components/forms/label-input.stories.ts:
I'd change this to be in the same storybook folder as the rest, for example in `Atoms/Forms/Label Input`

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+    let reader = new FileReader();
+    reader.readAsDataURL(image);
+    reader.onload = (e) => {
+      avatar = e.target.result;
```
```suggestion
  if (e.target?.result) {
        avatar = e.target.result.toString();
      }
```

My IDE nagged me about this because e.target might be null in error cases ;p

MRita443 replied:
Currently, this shows no error if the file uploaded is of the wrong type, we should add that

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+          if (e.key === 'Enter' || e.key === ' ') {
+            fileinput.click();
+          }
+        }}
```
These bindings are not working for me, are they for you?

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+      <img class="h-[200px] w-[200px] object-cover" src={avatar} alt="Avatar" />
+      <button
+        type="button"
+        aria-label="Change Avatar"
```
Because this is gonna be a reusable component, I'd recommend changing the aria-label and alts to use the `text` you receive or use more general terms like upload image :) For the same reason I'd also rename this to `image-input` and change references to `avatar`to `image` as well, but let me know what you think!

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+        type="button"
+        aria-label="Change Avatar"
+        class="absolute bottom-0 right-0 m-2 flex h-[15%] w-[15%] cursor-pointer items-center justify-center rounded-md bg-rose-950"
+        on:click={() => {
```
I think it would be more intuitive and more accessible on mobile if you could click anywhere on the square, not just on the edit button!

MRita443 commented on src/lib/components/forms/picture-input.svelte:
I'd also try to see if you can line up the text with the `*`, it looks a bit off-center currently!
![image](https://github.com/user-attachments/assets/031ade76-1d61-4fa2-8355-104b17753010)

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+        accept=".jpg, .jpeg, .png"
+        on:change={(e) => onFileSelected(e)}
+        bind:this={fileinput}
+      />
```
The button and input are repeated and can be moved to outside the if statement

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -0,0 +1,74 @@
+  import Icons from '$lib/components/icons/icons';
+
+  export let text: string;
+  let avatar, fileinput;
```
Don't forget the typescript types! :)
```suggestion
  let avatar: string;
  let fileinput: HTMLInputElement;
```

MRita443 commented on src/lib/components/forms/radio-buttons.svelte:
```diff
@@ -0,0 +1,29 @@
+      value={option}
+    />
+    <label
+      class="m-1 justify-self-start rounded-lg bg-taupe-200 px-5 py-1 text-rose-950"
+      for="radio-{option}"
+    >
```
```suggestion
      class="m-1 justify-self-start rounded-lg bg-taupe-200 px-5 py-1 text-rose-950"
      for="radio-{option}"
      class:selected={selectedOption === option}
    >
```

Maybe this would be a bit more readable! What do you think? Then we could remove the `radio-button` class and just do `  label.selected {...}` for the css

MRita443 commented on src/lib/components/forms/radio-buttons.svelte:
```diff
@@ -0,0 +1,29 @@
+  let selectedOption: string;
+</script>
+
+<div>
```
```suggestion
<fieldset>
```
I think this fits well

MRita443 commented on src/lib/components/forms/files-input.svelte:
This is missing Storybook stories, if you need any help let me know!

MRita443 commented on src/lib/components/forms/files-input.svelte:
```diff
@@ -0,0 +1,31 @@
+<script>
+  let files = null;
+
+  $: () => {
+    if (files) {
+      console.log(files);
+      for (const file of files) {
+        console.log(`${file.name}: ${file.size} bytes`);
+      }
+    }
+  };
```
Don't forget to remove when you're done :)

MRita443 commented on src/lib/components/forms/files-input.svelte:
```diff
@@ -0,0 +1,31 @@
+  {#if files}
+    {#each Array.from(files) as file}
+      <div class="rounded-lg bg-taupe-200 px-4 py-2 text-rose-950">
+        <p>{file.name}</p>
```
Hmm it would be useful if each file could have a little X to remove just that file, would that be doable?

MRita443 commented on src/lib/components/forms/files-input.svelte:
```diff
@@ -0,0 +1,31 @@
+    for="file-upload"
+    class="muted-red cursor-pointer rounded-lg px-4 py-2 text-white hover:bg-red-600"
+  >
+    Selecionar ficheiro
```
This should be an input to the component so that we can place whatever text here!

MRita443 commented on src/lib/components/forms/files-input.svelte:
```diff
@@ -0,0 +1,31 @@
+<div class="flex items-center space-x-4">
+  <label
+    for="file-upload"
+    class="muted-red cursor-pointer rounded-lg px-4 py-2 text-white hover:bg-red-600"
```
The button differs a bit currently from the mockups:
![image](https://github.com/user-attachments/assets/e5daeceb-f936-4f05-a6b1-a7cf736d595f)
We should add the red background color to the button and maybe make that color a bit darker on hover?

MRita443 reviewed (changes requested):
Great job! left some suggestions more about organization and edge cases, it's looking good tho

tomas-sucena committed:
refactor(radio button): replace 'div' with 'fieldset'

Co-authored-by: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
chore(picture input): add types

Co-authored-by: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
refactor(radio button): simplify CSS selector

Based on @MRita443 's suggestion, add class 'selected' to the label when hovered.

Co-Authored-By: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
chore(storybook): change directory structure

Co-Authored-By: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
refactor(picture input): reduce code inside 'if'-statement

tomas-sucena committed:
chore: format code

tomas-sucena committed:
refactor(radio button): simplify targeting selected labels

tomas-sucena commented:
@MRita443 Found an even better way to style selected labels on the `RadioButton` component. It requires neither a complex selector like the one I originally proposed nor JavaScript overhead like in your suggested change.

![image](https://github.com/user-attachments/assets/2d9bea91-5020-4ab4-9773-ea5d4a645f33)

What do you think?

MRita443 commented:
> @MRita443 Found an even better way to style selected labels on the `RadioButton` component. It requires neither a complex selector like the one I originally proposed nor JavaScript overhead like in your suggested change.
> 
> ![image](https://private-user-images.githubusercontent.com/93825634/393228968-2d9bea91-5020-4ab4-9773-ea5d4a645f33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNTIzNjIsIm5iZiI6MTczNjM1MjA2MiwicGF0aCI6Ii85MzgyNTYzNC8zOTMyMjg5NjgtMmQ5YmVhOTEtNTAyMC00YWI0LTk3NzMtZWE1ZDRhNjQ1ZjMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDE2MDEwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwNTg0ZDAzZjg4OWUzYjBhMmY3MzQ4OTBjYjM3MTMwMWZkNGVkYWQ2MmM4YTA1N2NkNTM5NGRhMWIyMWI1MmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3-jADTUgcu-d34wgT0Z_xre2LizVfKfXWOGfxdd-c2E)
> 
> What do you think?

Looks great! It's a lot cleaner than our previous options, and makes more sense imo.

tomas-sucena committed:
feat(radio button): add label

tomas-sucena committed:
refactor(radio button): add 'name' attribute

tomas-sucena committed:
fix(radio button): fix incorrect label binding

tomas-sucena committed:
feat(radio buttons): align legend with buttons

tomas-sucena committed:
refactor(picture input): code cleanup

tomas-sucena committed:
refactor(radio buttons): attempt to center legend

tomas-sucena commented:
@MRita443 Could you (or anyone who understands CSS😅) please help me center the legend with the radio buttons when 'horizontal' is set to `true`? I tried my best but right now it looks slightly off...

![image](https://github.com/user-attachments/assets/97e5cb94-5191-45d0-8666-d9ee666a3ec5)

After this, the radio buttons will finally be complete.

tomas-sucena committed:
feat(picture input): make image clickable

tomas-sucena committed:
feat(picture input): validate uploaded file

tomas-sucena committed:
feat(picture input): add popup for invalid images

tomas-sucena committed:
fix(picture input): reserve space for hidden button

Prevent the 'Remove image' button from pushing the picture display upwards when it appears.

tomas-sucena committed:
feat(picture input): add 'name' attribute

tomas-sucena committed:
chore: temporarily remove unfinished component

MRita443 commented on src/lib/components/forms/picture-input.svelte:
```diff
@@ -1,74 +1,68 @@
+    }}
+  >
+    {#if image}
+      <img class="h-[200px] w-[200px] rounded-md object-cover" src={image} alt="Avatar" />
```
```suggestion
      <img class="h-[200px] w-[200px] rounded-md object-cover" src={image} alt={`Selected ${name}`} />
```

MRita443 commented on src/lib/components/forms/picture-input.stories.ts:
```diff
@@ -0,0 +1,18 @@
+  },
+  parameters: {
+    layout: 'centered'
+  }
```
```suggestion
  parameters: {
    layout: 'centered',
    controls: { exclude: ['name'] }
  }
```

MRita443 commented on src/routes/(app)/_components/layout/notifications/notification-messages.ts:
```diff
@@ -1,5 +1,6 @@
 }>({
-  COPY_EMAIL: 'O email foi copiado para o teu clipboard :)'
+  COPY_EMAIL: 'O email foi copiado para o teu clipboard :)',
+  INVALID_IMAGE: 'Imagem inválida! Por favor, seleciona outra.'
```
I think this is more clear about the file type being the problem

```suggestion
   NOT_AN_IMAGE: 'O ficheiro não é uma imagem. Por favor, seleciona outro.'
```

MRita443 reviewed (changes requested):
It's looking pretty good! I left some small comments. Besides those, I think it would be nice to add the horizontal/vertical label option to `LabelInput` too, as well as allow for `RadioButton` to have no label.

And it would also be beneficial to test the functionality with storybook interaction tests

tomas-sucena committed:
fix(picture input): fix storybook

Exclude 'name' from the list of attributes that can be changed on the Storybook.

Co-authored-by: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
fix(picture input): improve error message

Co-Authored-By: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
refactor(profile picture): improve alternate text

Co-Authored-By: Rita Lopes <93883163+MRita443@users.noreply.github.com>

tomas-sucena committed:
feat: make form labels optional

tomas-sucena committed:
refactor(radio buttons): improve code readibility

tomas-sucena committed:
feat(label input): align label with textbox

MRita443 reviewed (approved):
Great job!! :) Will merge now for agility but we should add interaction tests for the new components
